File: m/rls/rlb/imexe.c
/* file -  imexe - execute image */
#include "m:\rid\rider.h"
#include "m:\rid\imdef.h"
#if Dos
#include "m:\rid\dsdef.h"
#include "m:\rid\dslib.h"
#elif  Win || Wnt
#include "m:\rid\stdef.h"
#include "m:\rid\mxdef.h"
#include <windows.h>
#endif 
#define STDOUT (2)
/* code -  im_exe - execute an image */
im_exe(
char *spc ,
char *cmd ,
int flg )
{ 
#if Dos
  return ( ds_ExePrg (spc, rem));
#elif  Win || Wnt
  STARTUPINFO sup = {0};
  PROCESS_INFORMATION inf = {0};
  char tmp [mxLIN];
  ULONG sts = 0;
  HANDLE opt = 0;
  HANDLE err = 0;
  char *ptr ;
  HANDLE han ;
  st_cop (spc, tmp);
  st_app (" ", tmp);
  st_app (cmd, tmp);
  if ( flg && (ptr = st_fnd (" > ", tmp)) != 0) {
    *ptr = 0;
    han = CreateFile (ptr+3, GENERIC_WRITE,FILE_SHARE_WRITE, 0,CREATE_ALWAYS, 0, 0);
    PUT("Handle=%X\n", han);
    opt = GetStdHandle (STD_OUTPUT_HANDLE);
    err = GetStdHandle (STD_ERROR_HANDLE);
    SetStdHandle (STD_OUTPUT_HANDLE, han);
    PUT("SetStdHandle=%d\n", (SetStdHandle (STD_ERROR_HANDLE, han)));
    CloseHandle (han);
  } 
  for(;;)  {
    if( (sts = CreateProcess (NULL, tmp,NULL, NULL, 1, 0, NULL, NULL, &sup, &inf)) == 0)break;
    WaitForSingleObject (inf.hProcess, INFINITE);
    GetExitCodeProcess (inf.hProcess, &sts);
   break;} 
  if ( opt) {SetStdHandle (STD_OUTPUT_HANDLE, opt) ;}
  if ( err) {SetStdHandle (STD_ERROR_HANDLE, err) ;}
  return ( sts);
#endif 
} 

File: m/rls/imexe.r
file	imexe - execute image
include	rid:rider
include	rid:imdef
include	rid:dbdef
include	rid:fidef
include	rid:stdef
include	rid:mxdef
include	rid:wimod

;	im_exe	Execute a program -- wait for completion
;	im_spn	Spawn a program -- do not wait
;
;	Status codes are application specific.
;	DOS/Windows rules are:
;	0	success
;	1	error level 1
;	2	error level 2
;	etc

	STDOUT := (2)

code	im_exe - execute an image

  func	im_exe
	spc : * char
	cmd : * char
	flg : int
  is
	sup : STARTUPINFO = {0}
	inf : PROCESS_INFORMATION = {0}
	tmp : [mxLIN] char
	sts : LONG = 0
	opt : HANDLE = 0
	err : HANDLE = 0
	ptr : * char
	han : HANDLE

;	st_cop (spc, tmp)
	fi_def (spc, ".exe", tmp)
	fi_loc (tmp, tmp)
	st_app (" ", tmp)
	st_app (cmd, tmp)
	if flg && (ptr = st_fnd (" > ", tmp)) ne 
	   *ptr = 0
	   han = CreateFile (ptr+3, GENERIC_WRITE,
	        FILE_SHARE_WRITE, 0,CREATE_ALWAYS, 0, 0)
PUT("Handle=%X\n", han)
	   opt = GetStdHandle (STD_OUTPUT_HANDLE)
	   err = GetStdHandle (STD_ERROR_HANDLE)
	   SetStdHandle (STD_OUTPUT_HANDLE, han)
	   SetStdHandle (STD_ERROR_HANDLE, han)
PUT("SetStdHandle=%d\n", that)
	   CloseHandle (han)
	end

      repeat
;;;PUT("tmp=[%s]\n", tmp)

	sts = CreateProcess (<>, tmp,
	 <>, <>, 1, 0, <>, <>, &sup, &inf)
	quit if fail
;;;	quit db_lst ("im_exe") if fail
	WaitForSingleObject (inf.hProcess, INFINITE)
;sic]	ignore fail
	GetExitCodeProcess (inf.hProcess, &sts)
      never
	SetStdHandle (STD_OUTPUT_HANDLE, opt) if opt
	SetStdHandle (STD_ERROR_HANDLE, err) if err
	reply sts
  end
